### PR TITLE
User Selectable Qt Version at Config time

### PR DESCRIPTION
**Pull request by sithlord48**
## Contributor Checklist:

* [X] I have created a file in the `doc/newsfragments` directory *IF* it is a
      user-visible change (and make sure to read the `README.md` in that directory) 

Fixes: #1395

Allow the user to select the Qt Version if both Qt5 and Qt6 are installed.
 - Set `QT_DEFAULT_MAJOR_VERSION` to 6 to build with Qt6
 - Set `QT_DEFAULT_MAJOR_VERSION` to 5 to build with Qt5 ( Default if not set)
 
Few Changes to the GUI cmakelist
 1. Use `QT_DEFAULT_MAJOR_VERSION` to control the version of Qt Required
          -  Qt5 Builds Require Qt Version 5.9+
          -  Qt6 Builds Require Qt Version 6.2+
 2. Use `QT_DEFAULT_MAJOR_VERSION` in place of `Qt#_Found` for tool guards

### CODE REVIEW

File: src/gui/CMakeLists.txt [hunk 1]
```diff
@@ -1,14 +1,20 @@
-cmake_minimum_required (VERSION 3.4)
```
Comment (shymega) [inline src/gui/CMakeLists.txt:1]: This should be in a separate PR.

File: src/gui/CMakeLists.txt [hunk 2]
```diff
@@ -173,7 +179,7 @@ set(TS_FILES
 )
 
 set_source_files_properties(${TS_FILES} PROPERTIES OUTPUT_LOCATION "${CMAKE_CURRENT_SOURCE_DIR}/res/lang")
-if (Qt5_FOUND)
+if(Qt5_FOUND)
```
Comment (shymega) [inline src/gui/CMakeLists.txt:182]: I don't understand why this was changed. Ideally, all changes should follow the existing coding style - including in CMake. In this case, we should keep a single space.

File: doc/newsfragments/qt-version-selection.txt [hunk 3]
```diff
@@ -0,0 +1 @@
+If your system has both Qt5 and Qt6 installed you can select the version that will be used by the build by setting the cmake Variable QT_DEFAULT_MAJOR_VERSION to either 5 or 6. If its not set Qt5 will be used.
```
Comment (shymega) [inline doc/newsfragments/qt-version-selection.txt:1]: I think the wording here could be improved.

It comes across as a bit confusing, and variables should be in backticks. Additionally, the file should not be `.txt`.

Finally, I think grammar could be adjusted; but that might be me nitpicking. I can change this later.

**Review by shymega — CHANGES_REQUESTED** [review]
Just a few nitpicks.

File: src/gui/CMakeLists.txt [hunk 4]
```diff
@@ -1,14 +1,20 @@
...
+endif()
+
+find_package (Qt${QT_DEFAULT_MAJOR_VERSION} ${REQUIRED_QT_VERSION} COMPONENTS Core Widgets Network LinguistTools REQUIRED)
 if (Qt6_FOUND)
     find_package (Qt6 REQUIRED COMPONENTS Core5Compat)
 endif()
 
-message(STATUS "Using Qt" ${QT_VERSION_MAJOR})
-
+message(STATUS "Using Qt ${QT_DEFAULT_MAJOR_VERSION}")
```
Comment (shymega) [inline src/gui/CMakeLists.txt:17]: I like the logic here. Thank you!

File: src/gui/CMakeLists.txt [hunk 5]
```diff
@@ -259,5 +265,5 @@ if(INPUTLEAP_BUILD_TESTS)
              WORKING_DIRECTORY ${CMAKE_RUNTIME_OUTPUT_DIRECTORY})
 
     target_include_directories(guiunittests PUBLIC ../../ext)
-    target_link_libraries(guiunittests gtest gmock Qt${QT_VERSION_MAJOR}::Core Qt${QT_VERSION_MAJOR}::Widgets Qt${QT_VERSION_MAJOR}::Network ${libs})
+    target_link_libraries(guiunittests gtest gmock Qt${QT_DEFAULT_MAJOR_VERSION}::Core Qt${QT_DEFAULT_MAJOR_VERSION}::Widgets Qt${QT_DEFAULT_MAJOR_VERSION}::Network ${libs})
```
Comment (shymega) [inline src/gui/CMakeLists.txt:268]: The logic side of things here is sound. Approving.

**Review by shymega — APPROVED** [review]
Other than my requested changes, this is a sound PR. However, I want to get @p12tic's eyes on it before merging.

File: src/gui/CMakeLists.txt (hunk 1 above)
Comment (sithlord48) [reply src/gui/CMakeLists.txt:1]: ok

File: doc/newsfragments/qt-version-selection.txt (hunk 3 above)
Comment (sithlord48) [reply doc/newsfragments/qt-version-selection.txt:1]: Do you have some better wording ? what should the extension be ?

File: src/gui/CMakeLists.txt (hunk 2 above)
Comment (sithlord48) [reply src/gui/CMakeLists.txt]: typo, fixed now

File: src/gui/CMakeLists.txt (hunk 1 above)
Comment (sithlord48) [reply src/gui/CMakeLists.txt:1]: Split in to https://github.com/input-leap/input-leap/pull/1714

File: doc/newsfragments/qt-version-selection.txt (hunk 3 above)
Comment (sithlord48) [reply doc/newsfragments/qt-version-selection.txt]: I have added to the Qt6.bugfix file this line 
`To Force use of Qt6 set the cmake variable QT_DEFAULT_MAJOR_VERSION to 6.`
I think this is the most sane place to inform builders.

Comment (shymega) [reply doc/newsfragments/qt-version-selection.txt]: I've edited it. This isn't a bug fix, so it's not a good place. It's a feature.

**Comment by shymega** [comment]
I've just force-pushed and rebased against `master`.

I've edited the commit message to explain the changes and provide background info. I've also signed off the commit.

The newsfragment was an addition to an existing one, which breaks history. I have created a new newsfragment as a 'feature' - this PR is a feature. I've also provided context on the change.

Finally, I rebased the PR against `master`, and pushed it to your branch, @sithlord48.

You will need to force pull with `git pull --rebase $REMOTE $BRANCH`.

Once @p12tic reviews, we can then merge.

**Comment by shymega** [comment]
@sithlord48 I can expedite this PR.

But, can we document how users set the Qt version? I would expect a CMake `option` in the root `CMakeLists.txt`.

Cheers.

**Comment by sithlord48** [comment]
> @sithlord48 I can expedite this PR.
> 
> But, can we document how users set the Qt version? I would expect a CMake `option` in the root `CMakeLists.txt`.
> 
> Cheers.
It is documented in the related news fragment. 
This is a variable that is part of the [Qt's Cmake](https://doc.qt.io/qt-6/cmake-variable-reference.html) setup not an option we created.

**Comment by shymega** [comment]
OK, I'm happy with this. Let's merge.
